File: 03-系统编程/03-内存管理/分配器/bump-allocator/c/bump_allocator.c
/*
 * bump_allocator.c — 最小 bump (arena) 分配器
 *
 * 工作机制:
 *   - 预分配一大块连续内存(mmap PROT_READ|PROT_WRITE MAP_PRIVATE|MAP_ANONYMOUS)
 *   - 维护一个 offset(从 base 起),每次分配把 offset 对齐到 alignment,
 *     然后对齐的地址返回给调用方、offset += size
 *   - 释放只能整块重置(offset = 0)— 不能 free 单个对象
 *
 * 对齐数学:
 *   current 是当前 offset 对应的绝对地址(uintptr_t)
 *   aligned = (current + alignment - 1) & ~(alignment - 1)
 *   即把 current 向上取整到 alignment 的倍数
 *
 * 注意事项:
 *   - struct 中不能放匿名 struct(必须 typedef struct Arena { ... } Arena;)
 *     以便内部 struct Arena *next 自引用
 *   - alignment 必须为 2 的幂(分配函数会校验),否则 & ~(alignment-1) 失效
 *   - 大对象请求 alignment + size 后可能超过 capacity;本 demo 演示单块,
 *     超过则返回 NULL
 */
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <unistd.h>

#define ARENA_CAPACITY (64 * 1024)  /* 64 KiB 单块 */

typedef struct Arena {
    char    *base;       /* mmap 出的整块首地址                 */
    size_t   capacity;   /* 总容量(字节)                      */
    size_t   offset;     /* 下一个分配的起点(offset 相对 base) */
} Arena;

/* 初始化:把 Arena 控制结构与底层 buffer 放在同一 mmap 区域,简化管理。       */
static Arena *arena_create(size_t capacity)
{
    /* Arena 控制头 + 数据段一并 mmap,只一次系统调用。                       */
    size_t total   = sizeof(Arena) + capacity;
    void  *region  = mmap(NULL, total,
                          PROT_READ | PROT_WRITE,
                          MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
    if (region == MAP_FAILED) {
        perror("mmap");
        return NULL;
    }
    Arena *a      = (Arena *)region;
    a->base       = (char *)region + sizeof(Arena);
    a->capacity   = capacity;
    a->offset     = 0;
    return a;
}

/* 整块重置:offset 回 0,后续分配复用同一段内存。                            */
static void arena_reset(Arena *a)
{
    a->offset = 0;
}

/* 销毁:把 mmap 区域还给内核。                                                 */
static int arena_destroy(Arena *a)
{
    if (!a) return 0;
    size_t total = sizeof(Arena) + a->capacity;
    return munmap(a, total);
}

/* 对齐到 2 的幂;返回 (x 向上取整到 alignment 倍数)。                       */
static inline uintptr_t align_up(uintptr_t x, size_t alignment)
{
    /* alignment 必为 2 的幂;否则 & ~(alignment-1) 会给出错误结果。          */
    return (x + alignment - 1) & ~(alignment - 1);
}

/* 分配 size 字节,按 alignment 对齐。                                          *
 * 返回:成功 → 对齐后的指针;失败 → NULL.                                       */
static void *arena_alloc(Arena *a, size_t size, size_t alignment)
{
    if (!a || alignment == 0 || (alignment & (alignment - 1)) != 0) {
        /* alignment 必须是 2 的幂,这是 (x + a-1) & ~(a-1) 能正确取整的前提 */
        fprintf(stderr, "arena_alloc: invalid alignment (%zu)\n", alignment);
        return NULL;
    }

    /* 1) 当前 offset 转绝对地址 → 向上对齐                                */
    uintptr_t current  = (uintptr_t)(a->base + a->offset);
    uintptr_t aligned  = align_up(current, alignment);
    size_t    pad      = (size_t)(aligned - current);   /* 填充字节数      */

    /* 2) 加上 size 检查是否超容量                                          */
    if (a->offset + pad + size > a->capacity) {
        fprintf(stderr, "arena_alloc: OOM (need %zu, free %zu)\n",
                pad + size, a->capacity - a->offset);
        return NULL;
    }

    /* 3) 提交新的 offset                                                  */
    a->offset += pad + size;
    return (void *)aligned;
}

/* ----------------------- demo ----------------------- */

typedef struct {
    const char *name;
    unsigned    id;
} Data;

static void demo_basic(Arena *a)
{
    printf("[1] basic allocations (alignment = sizeof(void*) = %zu)\n",
           sizeof(void *));

    int  *iptr  = arena_alloc(a, sizeof(int), _Alignof(int));
    *iptr = 42;
    printf("    int*  -> offset-aligned to %zu-byte boundary, value=%d\n",
           _Alignof(int), *iptr);

    double *dptr = arena_alloc(a, sizeof(double), _Alignof(double));
    *dptr = 3.14;
    printf("    double* -> value=%.2f\n", *dptr);

    Data  *d   = arena_alloc(a, sizeof(Data), _Alignof(Data));
    d->name = "hello";
    d->id   = 7;
    printf("    struct Data -> {name=\"%s\", id=%u}\n", d->name, d->id);

    printf("    arena offset after 3 allocs = %zu bytes\n", a->offset);
}

static void demo_alignment(Arena *a)
{
    printf("\n[2] explicit alignment (32-byte)\n");

    /* 先分配一个 1 字节对象,把 offset 推到非对齐位置;然后用 32 字节对齐查看填充  */
    char  *c    = arena_alloc(a, sizeof(char), 1);
    *c = 'X';
    size_t before = a->offset;

    int  *aligned32 = arena_alloc(a, sizeof(int), 32);
    size_t after    = a->offset;

    printf("    char*  @offset %zu   (before)\n", before - 1);
    printf("    int*   @alignment 32, padding = %zu bytes (offset went %zu -> %zu)\n",
           after - before - sizeof(int), before, after);
    printf("    int value = %d\n", *aligned32);
}

static void demo_reset_reuse(Arena *a)
{
    printf("\n[3] reset & reuse — bump allocator has no per-object free\n");

    size_t before = a->offset;
    arena_reset(a);
    printf("    reset(): offset %zu -> 0\n", before);

    /* 第二次分配能从 offset=0 重用同一段 buffer。                              */
    int *p1 = arena_alloc(a, sizeof(int), _Alignof(int));
    *p1 = 100;
    int *p2 = arena_alloc(a, sizeof(int), _Alignof(int));
    *p2 = 200;
    printf("    reuse: p1=%d, p2=%d  (offset=%zu)\n", *p1, *p2, a->offset);
    printf("    note: previous int=42, double=3.14, etc. are now logically freed\n");
}

static void demo_oob_check(Arena *a)
{
    printf("\n[4] OOM check — request too large for remaining capacity\n");
    size_t remaining = a->capacity - a->offset;
    printf("    remaining = %zu bytes, capacity = %zu bytes\n",
           remaining, a->capacity);

    size_t huge = a->capacity;  /* 故意请求 = capacity,显然超 remaining         */
    void *p = arena_alloc(a, huge, 1);
    if (!p) printf("    arena_alloc(%zu) returned NULL as expected ✓\n", huge);
    else    printf("    unexpected: alloc succeeded??\n");
}

int main(void)
{
    printf("=== bump (arena) allocator demo ===\n");
    printf("capacity = %d KiB\n", ARENA_CAPACITY / 1024);

    Arena *a = arena_create(ARENA_CAPACITY);
    if (!a) return 1;

    demo_basic(a);
    demo_alignment(a);
    demo_reset_reuse(a);
    demo_oob_check(a);

    arena_destroy(a);
    printf("\n[ok] arena destroyed.\n");
    return 0;
}

File: 03-系统编程/03-内存管理/分配器/bump-allocator/go/bump_allocator.go
// bump_allocator.go — 最小 bump (arena) 分配器(单块 Go 版)
//
// 工作机制与 C 版完全对应:
//   - 预分配一大块字节切片(make([]byte, capacity))
//   - 维护 a.offset(相对 &a.buf[0] 的字节数)
//   - 每次分配把 offset 对齐到 alignment,返回 &buf[aligned] 的切片;
//     offset += 填充 + size
//   - 释放只能整块重置(Reset),不能 free 单个对象
//
// 与 C 版的差异:
//   - 用 Go 内置 make/切片代替 mmap,无须手动管理底层页
//   - 用 unsafe.Offsetof / unsafe.Alignof 取类型对齐(等价 C 的 _Alignof(T))
//   - 用 reflect.Type 与 unsafe 算类型 size(代替 sizeof(T))
//
// 注意事项:
//   - alignment 必须为 2 的幂,与 C 版约束一致
//   - 返回的 []byte 指向 a.buf 内部;a.Reset() 之后该切片仍指向同样地址,
//     但语义上属于复用;调用方写入前必须重新拿切片
package main

import (
	"errors"
	"fmt"
	"reflect"
	"runtime"
	"unsafe"
)

// ---------- Arena 数据结构 ----------

type Arena struct {
	buf      []byte   // 底层 buffer
	offset   int      // 下一个分配起点(offset,相对 &buf[0])
	capacity int      // 总容量
}

// NewArena 创建一段容量为 capacity(必须 2 的幂)的 bump arena。
func NewArena(capacity int) (*Arena, error) {
	if capacity <= 0 || capacity&(capacity-1) != 0 {
		return nil, fmt.Errorf("capacity must be a positive power of 2, got %d", capacity)
	}
	buf := make([]byte, capacity)
	return &Arena{buf: buf, offset: 0, capacity: capacity}, nil
}

// Reset 把 offset 归零;老的切片在逻辑上视为失效。
func (a *Arena) Reset() { a.offset = 0 }

// alignUp 把 x 向上取整到 alignment 的倍数(alignment 必为 2 的幂)。
func alignUp(x, alignment uintptr) uintptr {
	return (x + alignment - 1) &^ (alignment - 1)
}

// Alloc 分配 size 字节,按 alignment 向上对齐;返回一段写入即写入底层 buffer 的切片。
//
// 与 C 版 strict 返回 void* 不同,这里返回 []byte 更符合 Go 习惯。
// 分配失败(对齐非法或 OOM)返回 error,nil。
func (a *Arena) Alloc(size, alignment int) ([]byte, error) {
	if alignment <= 0 || alignment&(alignment-1) != 0 {
		return nil, fmt.Errorf("alignment must be a positive power of 2, got %d", alignment)
	}

	// 1) 当前 offset 对应的"绝对地址"(用 uintptr 模拟 C 的指针 → 整数)
	current := uintptr(unsafe.Pointer(&a.buf[a.offset]))
	aligned := alignUp(current, uintptr(alignment))
	pad := int(aligned - current)

	// 2) 检查剩余容量
	if a.offset+pad+size > a.capacity {
		return nil, fmt.Errorf("arena OOM: need %d bytes, only %d bytes free",
			pad+size, a.capacity-a.offset)
	}

	// 3) 推进 offset 并返回切片(指向 a.buf 内部)
	a.offset += pad + size
	return a.buf[aligned : aligned+uintptr(size)], nil
}

// AlignedSize 返回类型 T 的对齐(用 unsafe.Alignof 取代 C 的 _Alignof(T))。
// 这是 Go 1.20+ 提供的 stdlib 工具,1.21 之前可用 reflect.Type.Align()。
func AlignedSize[T any]() (size, align int) {
	var zero T
	t := reflect.TypeOf(zero)
	return int(t.Size()), int(t.Align())
}

// ------------------------- demo -------------------------

type Data struct {
	Name string // 16-byte string header
	ID   uint32
}

func demoBasic(a *Arena) {
	fmt.Printf("[1] basic allocations (default align = %d bytes)\n", unsafe.Alignof(int(0)))

	// int
	intSize, intAlign := AlignedSize[int]()
	view, err := a.Alloc(intSize, intAlign)
	if err != nil {
		fmt.Println("    alloc err:", err)
		return
	}
	*(*int)(unsafe.Pointer(&view[0])) = 42
	fmt.Printf("    int*   -> %d-byte aligned, value=%d\n", intAlign, *(*int)(unsafe.Pointer(&view[0])))

	// float64(double)
	dSize, dAlign := AlignedSize[float64]()
	dv, _ := a.Alloc(dSize, dAlign)
	*(*float64)(unsafe.Pointer(&dv[0])) = 3.14
	fmt.Printf("    double* -> value=%.2f\n", *(*float64)(unsafe.Pointer(&dv[0])))

	// struct Data
	sSize, sAlign := AlignedSize[Data]()
	sv, _ := a.Alloc(sSize, sAlign)
	d := (*Data)(unsafe.Pointer(&sv[0]))
	d.Name = "hello"
	d.ID = 7
	fmt.Printf("    struct Data -> {Name=%q, ID=%d}\n", d.Name, d.ID)
	fmt.Printf("    arena offset after 3 allocs = %d bytes\n", a.offset)
}

func demoAlignment(a *Arena) {
	fmt.Println("\n[2] explicit alignment (32-byte)")

	// 先 alloc 1 字节 char,把 offset 推到非对齐位置
	a.Alloc(1, 1)
	before := a.offset

	v, _ := a.Alloc(4, 32)
	after := a.offset
	*(*int32)(unsafe.Pointer(&v[0])) = 7
	fmt.Printf("    char* offset before = %d\n", before-1)
	fmt.Printf("    int*  32-byte aligned; padding = %d bytes (offset %d -> %d)\n",
		after-before-4, before, after)
	fmt.Printf("    int value = %d\n", *(*int32)(unsafe.Pointer(&v[0])))
}

func demoResetReuse(a *Arena) {
	fmt.Println("\n[3] reset & reuse — bump allocator has no per-object free")

	before := a.offset
	a.Reset()
	fmt.Printf("    reset(): offset %d -> 0\n", before)

	p1, _ := a.Alloc(4, 4)
	*(*int32)(unsafe.Pointer(&p1[0])) = 100
	p2, _ := a.Alloc(4, 4)
	*(*int32)(unsafe.Pointer(&p2[0])) = 200
	fmt.Printf("    reuse: p1=%d, p2=%d  (offset=%d)\n",
		*(*int32)(unsafe.Pointer(&p1[0])), *(*int32)(unsafe.Pointer(&p2[0])), a.offset)
	fmt.Println("    note: previous int=42, double=3.14, etc. are now logically freed")
}

func demoOOBCheck(a *Arena) {
	fmt.Println("\n[4] OOM check — request too large for remaining capacity")
	remaining := a.capacity - a.offset
	huge := a.capacity
	fmt.Printf("    remaining = %d bytes, capacity = %d bytes\n", remaining, a.capacity)
	if _, err := a.Alloc(huge, 1); err != nil {
		fmt.Printf("    arena.Alloc(%d) returned error as expected ✓ (%v)\n", huge, err)
	} else {
		fmt.Println("    unexpected: alloc succeeded??")
	}
}

func main() {
	fmt.Println("=== bump (arena) allocator demo (Go) ===")
	const CAP = 64 * 1024
	fmt.Printf("capacity = %d KiB\n", CAP/1024)

	a, err := NewArena(CAP)
	if err != nil {
		fmt.Println("NewArena:", err)
		return
	}

	demoBasic(a)
	demoAlignment(a)
	demoResetReuse(a)
	demoOOBCheck(a)

	// 显式把 a 标 nil 让 GC 回收;runtime.GC 只是为了观察;实际演示非必需
	a = nil
	runtime.GC()
	fmt.Println("\n[ok] arena discarded (GC will free backing slice).")

	// 防止 errors 被 import-errcheck 之类工具误报
	_ = errors.New
}
